File: habit-tracker-backend/src/main/java/com/habittracker/service/HabitService.java
package com.habittracker.service;

import com.habittracker.dao.HabitDAO;
import com.habittracker.dto.HabitCreateDTO;
import com.habittracker.dto.HabitResponseDTO;
import com.habittracker.dto.HabitUpdateDTO;
import com.habittracker.model.Habit;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Servicio con la logica de negocio para la gestion de habitos.
 * Se encarga de crear, listar, actualizar y eliminar habitos logicamente.
 */
public class HabitService {

    private final HabitDAO habitDAO;

    public HabitService() {
        this.habitDAO = new HabitDAO();
    }

    /**
     * Crea un nuevo habito a partir de los datos del DTO.
     *
     * @param dto Datos del formulario de creacion
     * @return HabitResponseDTO con los datos del habito creado
     */
    public HabitResponseDTO createHabit(HabitCreateDTO dto) {
        Habit habito = new Habit();
        habito.setUserId(dto.getUserId());
        habito.setName(dto.getName());
        habito.setDescription(dto.getDescription());
        habito.setEmoji(dto.getEmoji());
        habito.setFrequency(dto.getFrequency());
        habito.setTargetDays(dto.getTargetDays());
        habito.setColor(dto.getColor());

        Habit habitoCreado = habitDAO.save(habito);

        if (habitoCreado == null) {
            throw new RuntimeException("Error al crear el habito");
        }

        return new HabitResponseDTO(habitoCreado);
    }

    /**
     * Obtiene todos los habitos activos de un usuario.
     *
     * @param userId ID del usuario
     * @return Lista de HabitResponseDTO
     */
    public List<HabitResponseDTO> getHabitsByUser(int userId) {
        List<Habit> habitos = habitDAO.findByUserId(userId);
        return habitos.stream()
                .map(HabitResponseDTO::new)
                .collect(Collectors.toList());
    }

    /**
     * Actualiza un habito existente.
     *
     * @param dto Datos actualizados del habito
     * @return HabitResponseDTO con los datos actualizados, o null si no existe
     */
    public HabitResponseDTO updateHabit(HabitUpdateDTO dto) {
        Habit habito = new Habit();
        habito.setId(dto.getId());
        habito.setName(dto.getName());
        habito.setDescription(dto.getDescription());
        habito.setEmoji(dto.getEmoji());
        habito.setFrequency(dto.getFrequency());
        habito.setTargetDays(dto.getTargetDays());
        habito.setColor(dto.getColor());

        boolean actualizado = habitDAO.update(habito);

        if (!actualizado) {
            return null;
        }

        return new HabitResponseDTO(habito);
    }

    /**
     * Elimina un habito logicamente (is_active = false).
     *
     * @param habitId ID del habito a eliminar
     * @return true si se elimino correctamente
     */
    public boolean deleteHabit(int habitId) {
        return habitDAO.deactivate(habitId);
    }
}

File: habit-tracker-frontend/src/app/services/habit.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Habit, HabitCreate, HabitUpdate } from '../models/habit.model';

/**
 * Servicio que centraliza las llamadas al backend para la gestion de habitos.
 * Proporciona metodos CRUD para crear, listar, actualizar y eliminar habitos.
 */
@Injectable({
  providedIn: 'root'
})
export class HabitService {
  private http = inject(HttpClient);
  private apiUrl = 'http://localhost:8080/habit-tracker-backend/api/habits';

  /**
   * Obtiene la lista de habitos activos de un usuario.
   * @param userId - El id del usuario
   * @returns Observable con un array de Habit
   */
  getHabitsByUser(userId: number): Observable<Habit[]> {
    return this.http.get<Habit[]>(`${this.apiUrl}?userId=${userId}`);
  }

  /**
   * Crea un nuevo habito.
   * @param habit - Los datos del habito a crear
   * @returns Observable con el habito creado
   */
  createHabit(habit: HabitCreate): Observable<Habit> {
    return this.http.post<Habit>(this.apiUrl, habit);
  }

  /**
   * Actualiza un habito existente.
   * @param id - El id del habito a actualizar
   * @param habit - Los datos actualizados del habito
   * @returns Observable con el habito actualizado
   */
  updateHabit(id: number, habit: HabitUpdate): Observable<Habit> {
    return this.http.put<Habit>(`${this.apiUrl}/${id}`, habit);
  }

  /**
   * Elimina (borrado logico) un habito.
   * @param id - El id del habito a eliminar
   * @returns Observable con un mensaje de confirmacion
   */
  deleteHabit(id: number): Observable<{ mensaje: string }> {
    return this.http.delete<{ mensaje: string }>(`${this.apiUrl}/${id}`);
  }
}
